packages/indicators/src/AnalyticsRepository.ts

```diff
@@ -0,0 +1,128 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2020 Beyond Essential Systems Pty Ltd
+ */
+
+import groupBy from 'lodash.groupby';
+
+import { Aggregator, expandFetchOptionsToSpanAggregations } from '@tupaia/aggregator';
+import { comparePeriods, dateStringToPeriod } from '@tupaia/utils';
+import { Aggregation, Analytic, FetchOptions } from './types';
+
+interface Data {
+  analyticsByElement: Record<string, Analytic[]>;
+  fetchOptions: FetchOptions;
+  expandedFetchOptions: FetchOptions;
+}
+
+/**
+ * This class encapsulates analytics fetching and retrieval of fetched data. The motivation
+ * behind it is to optimise data fetches by minimising their count.
+ *
+ * An example:
+ * * Indicator A wants to fetch FINAL_EACH_YEAR data for element BCD
+ * * Indicator B wants to fetch FINAL_EACH_YEAR for element BCD and a year offset of -1
+ * ("get me the results for the previous year")
+ * * The original date range is 2018 - 2020
+ *
+ * We can combine both fetches in one, and fetch BCD for years 2017 - 2020, so that we can include
+ * -1 offset data for the 2018 boundary.
+ *
+ * This means that the "scope" of the fetched data will be wider than the the original.
+ * This class hides the scope transformation and allows each client (indicator) to retrieve
+ * correct data based on the original scope (`fetchOptions`).
+ */
+export class AnalyticsRepository {
+  private aggregator: Aggregator;
+
+  private populatedData: Data | null = null;
+
+  constructor(aggregator: Aggregator) {
+    this.aggregator = aggregator;
+  }
+
+  private get data(): Data {
+    if (!this.populatedData) {
+      throw new Error('Please run fetchAnalytics first!');
+    }
+    return this.populatedData;
+  }
+
+  private set data(data: Data) {
+    this.populatedData = data;
+  }
+
+  isPopulated() {
+    return !!this.populatedData;
+  }
+
+  /**
+   * @param allAggregations A combined list of aggregations across all data elements
+   */
+  async populate(
+    dataElementCodes: string[],
+    allAggregations: Aggregation[],
+    fetchOptions: FetchOptions,
+  ) {
+    const expandedFetchOptions = expandFetchOptionsToSpanAggregations(
+      fetchOptions,
+      allAggregations,
+    );
+    const { results: analytics } = await this.aggregator.fetchAnalytics(
+      dataElementCodes,
+      expandedFetchOptions,
+    );
+
+    this.data = {
+      analyticsByElement: groupBy(analytics, 'dataElement'),
+      fetchOptions,
+      expandedFetchOptions,
+    };
```

Comment:
I would tend to store each of these three as individual fields - later it confuses me a bit to see `this.data.fetchOptions` etc.

Reply:
Of course you are right! Naming a variable `data` is definitely a smell. Will change

Reply:
I realise now why I stored all of them under a `data` field: because we need to check that the repository is populated before attempt to access any of those fields.

I could still add them as separate fields with a private getter/setter for each, but it's more verbose and it makes the distinction between populated/non-populated fields less clear. 

Based on the above, I'd prefer this to stay as is - I don't know if it's defined as such but in my head I'd call it the "Repository" pattern.

Reply:
Right....what about renaming the field to something like `populatedFields`?

Reply:
I guess fields is more specific than  "data"

Reply:
And less easily confused with the actual analytic data (which is the mistake I made when reading through the first time).

Thanks!

Reply:
I did the rename in [da4a989](https://github.com/beyondessential/tupaia/pull/1873/commits/da4a98914c7f424d8879ebb5d5e225ad4fd0132d): essentially I renamed `data` to `fields`.

Field access is done like this:
```ts
this.fields.fetchOptions
```

You may had the following in mind:
```ts
this.populatedFields.fetchOptions
```

But then I wouldn't know how to name the actual field in the class. Currently it is
```ts
private populatedFields: Fields | null = null;
```

and accessed via a getter/setter.

Happy to further rename it if you have another idea!

Reply:
No all good! You're right about my expectation but I saw that and realised why and am very happy with `fields` :-)